flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannelTest.java

```diff
@@ -301,81 +306,388 @@ public void testProducerFailedException() throws Exception {
+					for (Buffer buffer : floatingBuffers) {
+						buffer.recycle();
+					}
+
+					return null;
+				}
+			};
+
+			// Submit tasks and wait to finish
+			final List<Future<Void>> results = Lists.newArrayListWithCapacity(2);
+			results.add(executor.submit(requestBufferTask));
+			results.add(executor.submit(recycleBufferTask));
+			for (Future<Void> result : results) {
+				result.get();
+			}
+
+			final int numRequiredBuffers = backlog + numExclusiveSegments;
+			assertEquals("There should be " + numRequiredBuffers +" buffers available in channel.",
+				numRequiredBuffers, inputChannel.getNumberOfAvailableBuffers());
+			assertEquals("There should be no buffers available in buffer pool.",
+				0, bufferPool.getNumberOfAvailableMemorySegments());
+
+		} finally {
+			// Release all the buffer resources
+			inputChannel.releaseAllResources();
+
+			networkBufferPool.destroyAllBufferPools();
+			networkBufferPool.destroy();
+
+			executor.shutdown();
 		}
-		inputChannel.assignExclusiveSegments(exclusiveSegments);
+	}
 
-		assertEquals("The number of available buffers is not equal to the assigned amount.",
-			numExclusiveBuffers, inputChannel.getNumberOfAvailableBuffers());
+	/**
+	 * Tests to verify that there is no race condition with two things running in parallel:
+	 * recycling the exclusive or floating buffers and some other thread releasing the
+	 * input channel.
+	 */
+	@Test
+	public void testConcurrentRecycleAndRelease() throws Exception {
+		// Setup
+		final NetworkBufferPool networkBufferPool = new NetworkBufferPool(256, 32, MemoryType.HEAP);
+		final ExecutorService executor = Executors.newFixedThreadPool(2);
+		final SingleInputGate inputGate = createSingleInputGate();
+		final RemoteInputChannel inputChannel  = createRemoteInputChannel(inputGate);
+		inputGate.setInputChannel(inputChannel.partitionId.getPartitionId(), inputChannel);
+		try {
+			final int numFloatingBuffers = 128;
+			final int numExclusiveSegments = 2;
+			final BufferPool bufferPool = networkBufferPool.createBufferPool(numFloatingBuffers, numFloatingBuffers);
+			inputGate.setBufferPool(bufferPool);
+			inputGate.assignExclusiveSegments(networkBufferPool, numExclusiveSegments);
+
+			// Exhaust all the floating buffers
+			final List<Buffer> floatingBuffers = new ArrayList<>(numFloatingBuffers);
+			for (int i = 0; i < numFloatingBuffers; i++) {
+				Buffer buffer = bufferPool.requestBuffer();
+				assertNotNull(buffer);
+				floatingBuffers.add(buffer);
+			}
+
+			// Exhaust all the exclusive buffers
+			final List<Buffer> exclusiveBuffers = new ArrayList<>(numExclusiveSegments);
+			for (int i = 0; i < numExclusiveSegments; i++) {
+				Buffer buffer = inputChannel.requestBuffer();
+				assertNotNull(buffer);
+				exclusiveBuffers.add(buffer);
+			}
+
+			final Callable<Void> recycleBufferTask = new Callable<Void>() {
+				@Override
+				public Void call() throws Exception {
+					// Recycle all the exclusive buffers
```

Comment:
same here - two recycle threads?
Also, this method shares a lot of code with the previous test - can you extract the common parts into a shared method? (with only the non-recycling task being different and the checks after the test is run)

Reply:
I was actually hoping we could extract more into a common test method but it's probably best as you implemented it to keep the actual tests easier to understand